lib/aws/request.js

```diff
@@ -0,0 +1,218 @@
+    service = { name: service };
+  }
+  const BASE_BACKOFF = 5000;
+  const persistentRequest = async (f, numTry = 0) => {
+    try {
+      return await f();
+    } catch (e) {
+      const { providerError } = e;
+      if (
+        numTry < MAX_RETRIES &&
+        providerError &&
+        ((providerError.retryable &&
+          providerError.statusCode !== 403 &&
+          providerError.code !== 'CredentialsError') ||
+          providerError.statusCode === 429)
+      ) {
+        const nextTryNum = numTry + 1;
+        const jitter = Math.random() * 3000 - 1000;
+        // backoff is between 4 and 7 seconds
+        const backOff = BASE_BACKOFF + jitter;
+        log(
+          [
+            `Recoverable error occurred (${e.message}), sleeping for ~${Math.round(
+              backOff / 1000
+            )} seconds.`,
+            `Try ${nextTryNum} of ${MAX_RETRIES}`,
+          ].join(' ')
+        );
+        await wait(backOff);
+        return persistentRequest(f, nextTryNum);
+      }
+      throw e;
+    }
+  };
+  const request = await requestQueue.add(() =>
+    persistentRequest(async () => {
+      const awsService = getServiceInstance(service, method);
+      const req = awsService[method](...args);
+      try {
+        return await req.promise();
+      } catch (err) {
+        let message = err.message != null ? err.message : String(err.code);
+        if (message.startsWith('Missing credentials in config')) {
+          // Credentials error
+          // If failed at last resort (EC2 Metadata check) expose a meaningful error
+          // with link to AWS documentation
+          // Otherwise, it's likely that user relied on some AWS creds, which appeared not correct
+          // therefore expose an AWS message directly
+          let bottomError = err;
+          while (bottomError.originalError && !bottomError.message.startsWith('EC2 Metadata')) {
+            bottomError = bottomError.originalError;
+          }
+
+          const errorMessage = bottomError.message.startsWith('EC2 Metadata')
+            ? [
+                'AWS provider credentials not found.',
+                ' Learn how to set up AWS provider credentials',
+                ` in our docs here: <${chalk.green('http://slss.io/aws-creds-setup')}>.`,
+              ].join('')
+            : bottomError.message;
+          message = errorMessage;
+          // We do not want to trigger the retry mechanism for credential errors
+          throw Object.assign(new ServerlessError(errorMessage), {
+            providerError: Object.assign({}, err, { retryable: false }),
+          });
+        }
+        throw Object.assign(new ServerlessError(message, err.code), {
+          providerError: err,
+        });
+      }
+    })
+  );
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  promise: true,
+  normalizer: ([service, method, args]) => {
+    if (_.isString(service)) service = { name: service };
```

Comment:
It would be good to follow same resolution as in memoized function (and also we do not use lodash if not justified, e.g. in this case it could be `typeof service === 'string'`).

Anyway if we want to reflect what happens in `request`. It'll be:

```javascript
if (!isObject(service)) service = { name: ensureString(service) }
```

Reply:
I'll change it accordingly, I wanted to avoid having an error in the memoized part and keep it for the actual call.

Reply:
It's totally fine to reject here. It'll have exactly same effect